### ISSUE TITLE
Looking for some feedback on how to better handle different runtimes under the same build

### ISSUE THREAD

**Issue by paulormart** [paritytech/subxt#387]
Hi there,

This is not an issue. I'm just trying to get some help and chase up some feedback to better implement the projects where i'm using the latest version of this library `subxt`.

With the current api I find myself repeating a bunch of the same logic just to target different runtimes. I’m trying to find a better way of abstracting this if possible. 

As an example, let’s say I’m implementing a `cli` tool to target _Westend_, _Kusama_ and _Polkadot_ where the runtime chosen depends on some flag defined by the user of the tool. The tool should run exactly the same tasks no matter which runtime is specified.

Below is just a pseudo code of how i'm managing this and where i'm asking for some opinion on it.

Any feedback will be awesome. Thank you!

```

#[subxt::subxt(
    runtime_metadata_path = "metadata/westend_metadata.scale",
    generated_type_derives = "Clone, Debug"
)]
mod westend {}

#[subxt::subxt(
  runtime_metadata_path = "metadata/kusama_metadata.scale",
  generated_type_derives = "Clone, Debug"
)]
mod kusama {}

#[subxt::subxt(
  runtime_metadata_path = "metadata/polkadot_metadata.scale",
  generated_type_derives = "Clone, Debug"
)]
mod polkadot {}

pub type WestendApi = westend::RuntimeApi<DefaultConfig, DefaultExtra<DefaultConfig>>;
pub type KusamaApi = kusama::RuntimeApi<DefaultConfig, DefaultExtra<DefaultConfig>>;
pub type PolkadotApi = polkadot::RuntimeApi<DefaultConfig, DefaultExtra<DefaultConfig>>;

pub struct Context {
  pub client: Client<DefaultConfig>,,
}

impl Context {
  pub fn new(&self, url: &str) -> Context {

    let response = ClientBuilder::new()
      .set_url(url)
      .build::<DefaultConfig>()
      .await

    match response {
      Ok(client) => {
        Context {
          client: client
        }
      }
      Err(e) => {
        error!("{}", e)
      }
    }
  }
}

async fn main() -> Result<(), Error> {
    
    // Note: _runtime_
    // can be [Westend, Kusama, Polkadot] provided by a cli flag for example

    match runtime {
      "Polkadot" => {
        let ctx = Context::new("wss://api.polkadot.io");
        let api = ctx.client.to_runtime_api::<PolkadotApi>();
        run_job_on_polkadot(&api).await
      }
      "Kusama" => {
        let ctx = Context::new("wss://kusama-api.polkadot.io");
        let api = ctx.client.to_runtime_api::<KusamaApi>();
        run_job_on_kusama(&api).await
      },
      "Westend" => {
        let ctx = Context::new("wss://westend-api.polkadot.io");
        let api = ctx.client.to_runtime_api::<WestendApi>();
        run_job_on_westend(&api).await
      },
    }

    Ok(())
}

// The next 3 functions implement the same logic for each different network.
// 
// It will be nice if we could have only one function with the same logic for all 
// different runtimes. I just didn't find a way yet on how to abstratct the API..
// 
// How would you suggest to handle this in a better way? Any suggestion on how to improve this piece of code ?
// 
async fn run_job_on_polkadot(&api: PolkadotApi) -> Result<(), Error> {

  let active_validators = api.storage().session().validators(None).await?;
  let queued_keys = api.storage().session().queued_keys(None).await?;

  // .. any other logic here
}

async fn run_job_on_kusama(&api: KusamaApi) -> Result<(), Error> {

  let active_validators = api.storage().session().validators(None).await?;
  let queued_keys = api.storage().session().queued_keys(None).await?;

  // .. any other logic here
}

async fn run_job_on_westend(&api: WestendApi) -> Result<(), Error> {

    let active_validators = api.storage().session().validators(None).await?;
    let queued_keys = api.storage().session().queued_keys(None).await?;

    // .. any other logic here
}

```

**Comment by jsdw** [comment]
TL;DR I don't think you can simplify that logic using `subxt`.

This is an interesting question! As you've seen, `subxt` is focused on providing a compile-time API to target a given node. Because, in your example, 3 lots of types are generated for each runtime, even if they are identical there is currently no way to say "this function will accept any of these types so long as they both have some common functionality".

Now, Traits are the normal way to encode this in Rust, but for those to be of any use, `subxt` would have to be able to generate the code for multiple runtimes at once, find overlapping functionality, and generate one or more traits which describe the overlapping functionality. If that were the case, you could have calls like:

```
async fn run_job(&api: impl RuntimeHasSomeFunctionality) -> Result<(), Error> {

  let active_validators = api.storage().session().validators(None).await?;
  let queued_keys = api.storage().session().queued_keys(None).await?;

  // .. any other logic here
}
```

An alternative would be to wrap the generated runtimes into an enum yourself, so you'd have 1 function instead of 3, but you'd duplicate all of the code inside the match arms.

Perhaps in the future `subxt` could find this overlap and generate the appropraite traits/enums automatically for the "common" functionality, but for now your only way to avoid the code duplication would be to use a tool which can work out the interface of a node and encode/decode as needed at _runtime_ and not _compile time_. https://docs.rs/desub-current/0.1.0/desub_current/ is focused on _decoding_ things at runtime based on some node metadata, but I'm not aware of tooling to help with the runtime encoding story too offhand (though I'd very much like that to exist, and am keen on working in that direction).

**Comment by jsdw** [comment]
I'll close this for now, as I don't think that there are any short term plans to support this via subxt (rather, we should strive to provide the tooling to interact with nodes more dynamically, as opposed to subxt which is very compile-time focused) but please re-open if necessary :)

**Comment by paulormart** [comment]
Hi @jsdw many thanks for the feedback!

You are right, I'm really looking forward to see `subxt` supporting  something more dynamic in the future.

For now is working great, so i'm happy with the current API :)

Thank you
